torch/_numpy/_funcs_impl.py

```diff
@@ -360,11 +360,11 @@ def arange(
     # the dtype of the result
     if dtype is None:
         dtype = _dtypes_impl.default_dtypes().int_dtype
-    # XXX: default values do not get normalized
-    start, stop, step = (_util._coerce_to_tensor(x) for x in (start, stop, step))
-
-    dummy = torch.empty(1, dtype=dtype)
-    target_dtype = _dtypes_impl.result_type_impl(start, stop, step, dummy)
+        # XXX: default values do not get normalized
+        start, stop, step = (_util._coerce_to_tensor(x) for x in (start, stop, step))
+        target_dtype = _dtypes_impl.result_type_impl(start, stop, step)
+    else:
+        target_dtype = dtype
```

Comment:
What happens in NumPy if you pass `dtype=int8` but you pass a complex dtype for one or some of the parameters?

Reply:
Both numpy and we error out:

```
In [14]: np.arange(1j, 5j, 1j, dtype=np.int8)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[14], line 1
----> 1 np.arange(1j, 5j, 1j, dtype=np.int8)

TypeError: can't convert complex to float

In [15]: tnp.arange(1j, 5j, 1j, dtype=tnp.int8)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[15], line 1
----> 1 tnp.arange(1j, 5j, 1j, dtype=tnp.int8)

File ~/repos/pytorch/torch/_numpy/_normalizations.py:222, in normalizer.<locals>.normalizer_inner.<locals>.wrapped(*args, **kwds)
    209     args = (
    210         tuple(
    211             maybe_normalize(arg, parm)
   (...)
    214         + args[len(params.values()) :]
    215     )
    217 kwds = {
    218     name: maybe_normalize(arg, params[name]) if name in params else arg
    219     for name, arg in kwds.items()
    220 }
--> 222 result = func(*args, **kwds)
    224 # keepdims
    225 bound_args = None

File ~/repos/pytorch/torch/_numpy/_funcs_impl.py:372, in arange(start, stop, step, dtype, like)
    369 # work around RuntimeError: "arange_cpu" not implemented for 'ComplexFloat'
    370 work_dtype = torch.float64 if target_dtype.is_complex else target_dtype
--> 372 if (step > 0 and start > stop) or (step < 0 and start < stop):
    373     # empty range
    374     return torch.empty(0, dtype=target_dtype)
    376 result = torch.arange(start, stop, step, dtype=work_dtype)

RuntimeError: "gt_cpu" not implemented for 'ComplexDouble'

```

Reply:
same question but with float scalars and integer dtype. My point is, what happens if we use a dtype of a lower kind than the one we would infer normally. And what happens if it's of a lower width? Does numpy error, warning or comply?

Reply:
Numpy seems to comply to to whatever explicit dtype is given, which is what this PR does

```
(Pdb) np.arange(125, np.float64(130), dtype=np.int8)
array([ 125,  126,  127, -128, -127], dtype=int8)
(Pdb) np.arange(np.float64(5), dtype=np.float32)
array([0., 1., 2., 3., 4.], dtype=float32)
```

Reply:
Have some crazy edge-cases:
```python
>>> import numpy as np
>>> np.arange(0., 3., .5)
array([0. , 0.5, 1. , 1.5, 2. , 2.5])
>>> np.arange(0., 3., .5, dtype=np.int8)
array([0, 0, 0, 0, 0, 0], dtype=int8)
>>> import torch
to>>> torch.arange(0., 3., .5, dtype=torch.int8)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: step must be nonzero
```

I found this issue. Both NumPy and PyTorch compute the size of the output with the elements before casting (ok) and then  cast the inputs to the kind of `dtype` (this gives weird like in `np.arange(0., 3., 1.5, dtype=np.int8)`), but PyTorch errors out if the step is zero (which makes sense), while NumPy just goes on (weird).

This means that we need to cast the step to the correct kind and, if it's zero, return a tensor of `start`s and the correct size via `torch.full`.

All the other cases with `dtype` should be equivalent in PyTorch than in NumPy I believe.

Reply:
Now that I think about it, this is an issue when you mix it with the fact that we have to deal with 0-D tensors. In this case, just do this when the inputs are scalar and leave a note saying that we deviate slightly in this tiny edge-case.

Reply:
Now that scalars and casting are involved, I would seriously consider waiting out for numpy 2.0 to finalize their NEP 50 implementation + the overall 2.0 overhaul. The glaring error of arange(5, dtype=f32) -> f64 we can fix today.

Reply:
We can, setting the dtype to be `float64` (better, `default_float_dtype`)  if the output dtype is fp